### ISSUE TITLE
commit 9223315 ISSUE: Unpredictable and wrong behavior of openmp offload application on AMD GPU

### ISSUE THREAD

**Issue by fel-cab** [llvm/llvm-project#59341]
Application behaves unpredictable after commit 92233159035d1b50face95d886901cf99035bd99.
The application is SPEChpc tealeaf compiled for openmp offload.
CPU: AMD EPYC 7A53
GPU: AMD MI250X
Compilation Flags: -fopenmp -Wall -Ofast -foffload-lto -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx90a

The problem appears in different iterations of the execution, for different executions.
I have simplified as much as I could the application and still get the problem and added some printf to show the unexpected behavior. As an example, in one of the executions the error appears on iteration 175, where the value of the variable **rrn** shouldn't jump to the value on iteration 175:

`iteration 173, error: 1.563668706240932e+08   pw = 8.489109156989025e+12, rrn = 1.530719718661226e+08`
`iteration 174, error: 1.530719718661226e+08   pw = 8.237148061653313e+12, rrn = 1.521889487492852e+08`
`iteration 175, error: 1.521889487492852e+08   pw = 8.261711869147320e+12, rrn = 3.903098816183680e+299`

In a different execution the **rrn** value can get a wrong value on iteration 38 instead.

`iteration 36, error: 3.821804208275762e+09   pw = 2.165156939490312e+14, rrn = 3.561339554743209e+09`
`iteration 37, error: 3.561339554743209e+09   pw = 1.907022120691036e+14, rrn = 3.446392781811089e+09`
`iteration 38, error: 3.446392781811089e+09   pw = 1.945475027719899e+14, rrn = -1.080323384166538e+85`

If the code is compiled with the previous commit (a8ec170e01dd0d0cbf8c059aa7addaaf3d886876) the application doesn't presents this unpredictable behavior and all the executions are the same (as it should be).

The problem seems to be in one or both of functions `cg_calc_ur` or `cg_calc_w`  from the `cg.c` file.

I'm attaching the application code, with an compilation script, and the IR with the "good" and the "bad" commit obtained by compiling with `-Rpass=openmp-opt -Rpass-missed=openmp-opt -Rpass-analysis=openmp-opt -save-temps` and `opt -S -O3`.

[simplifiedTealeaf.zip](https://github.com/llvm/llvm-project/files/10155376/simplifiedTealeaf.zip)

**Comment by jhuber6** [comment]
I ran `./manualMake` and then `./tealeaf`. I didn't see any large jumps after over `1000` iterations. This was on a `gfx1030`. I don't have immediate access to a `gfx90a` but I can check on a `gfx908`.
```
iteration 1000, error: 7.511550781037347e+04   pw = 4.045969763662181e+09, rrn = 7.426966390083451e+04
iteration 1001, error: 7.426966390083451e+04   pw = 4.011078613947757e+09, rrn = 7.370572850763235e+04
iteration 1002, error: 7.370572850763235e+04   pw = 3.966062312764136e+09, rrn = 7.312464648380337e+04
iteration 1003, error: 7.312464648380337e+04   pw = 3.951308594701820e+09, rrn = 7.233031547645117e+04
iteration 1004, error: 7.233031547645117e+04   pw = 3.892159886348904e+09, rrn = 7.195283991595508e+04
iteration 1005, error: 7.195283991595508e+04   pw = 3.885798067466208e+09, rrn = 7.105352725748711e+04
iteration 1006, error: 7.105352725748711e+04   pw = 3.827298321941283e+09, rrn = 7.070973198488238e+04
iteration 1007, error: 7.070973198488238e+04   pw = 3.813745454087739e+09, rrn = 6.989101544576997e+04
iteration 1008, error: 6.989101544576997e+04   pw = 3.769876020102121e+09, rrn = 6.940740484350422e+04
iteration 1009, error: 6.940740484350422e+04   pw = 3.738778127126651e+09, rrn = 6.880477542354151e+04
iteration 1010, error: 6.880477542354151e+04   pw = 3.714873910341215e+09, rrn = 6.810339604179989e+04
iteration 1011, error: 6.810339604179989e+04   pw = 3.666568967098753e+09, rrn = 6.772752516839035e+04
iteration 1012, error: 6.772752516839035e+04   pw = 3.656829618398658e+09, rrn = 6.686372718969823e+04
iteration 1013, error: 6.686372718969823e+04   pw = 3.601488807459757e+09, rrn = 6.660206962470019e+04
iteration 1014, error: 6.660206962470019e+04   pw = 3.592878282025984e+09, rrn = 6.572552798166120e+04
iteration 1015, error: 6.572552798166120e+04   pw = 3.544329842993594e+09, rrn = 6.541340595941945e+04
```

My guess is this exposes a bug in the optimizations when the `IsSPMDMode` variable can be optimized out in LTO mode which the offending patch allows. Out of curiosity, does preventing this optimization hide the error?
```
diff --git a/openmp/libomptarget/DeviceRTL/src/Utils.cpp b/openmp/libomptarget/DeviceRTL/src/Utils.cpp
index e4920ab90450..13f36eadda77 100644
--- a/openmp/libomptarget/DeviceRTL/src/Utils.cpp
+++ b/openmp/libomptarget/DeviceRTL/src/Utils.cpp
@@ -19,7 +19,7 @@
 
 using namespace _OMP;
 
-extern "C" __attribute__((weak)) int IsSPMDMode;
+extern "C" __attribute__((weak, used)) int IsSPMDMode;
 
 /// Helper to keep code alive without introducing a performance penalty.
 extern "C" __attribute__((weak, optnone, cold, used, retain)) void
```

**Comment by fel-cab** [comment]
I compiled and executed with this patch and it does not fix this optimization error on my side.

**Comment by jhuber6** [comment]
That's interesting, how about if you add `optnone` as well?

**Comment by fel-cab** [comment]
I get the same result, the application keeps behaving strange.

**Comment by jhuber6** [comment]
> I get the same result, the application keeps behaving strange.

That's odd, because the patch you listed as the cause of this primarily removes those two attributes from the variable. So I'm unsure why it still exhibits this behavior with them added back in.

**Comment by fel-cab** [comment]
I tried something, that I'm not sure it makes sense. I left Utils.cpp as you have it in the commit, and added

```
diff --git a/openmp/libomptarget/DeviceRTL/src/Mapping.cpp b/openmp/libomptarget/DeviceRTL/src/Mapping.cpp
index f05e716b1607..8166f23a25d6 100644
--- a/openmp/libomptarget/DeviceRTL/src/Mapping.cpp
+++ b/openmp/libomptarget/DeviceRTL/src/Mapping.cpp
@@ -276,7 +276,7 @@ uint32_t mapping::getNumberOfProcessorElements() {
 
 // TODO: This is a workaround for initialization coming from kernels outside of
 //       the TU. We will need to solve this more correctly in the future.
-int __attribute__((weak)) SHARED(IsSPMDMode);
+int __attribute__((weak, used, retain)) SHARED(IsSPMDMode);
 
 void mapping::init(bool IsSPMD) {
   if (mapping::isInitialThreadInLevel0(IsSPMD))
```

Which, I think, it leaves Mapping.cpp as it was before your commit.

With this, the application runs correctly

**Comment by jdoerfert** [comment]
I think this is the same reduction issue we see with QMCPACK and similar benchmarks. We even have narrowed it down to the simplest league reduction you can think of. Where the race is exactly is still unknown though.

**Comment by fel-cab** [comment]
Yes, I'm still getting the same issue. I used a build from yesterday.
